pkg/controllers/statusaggregator/plugins/deployment.go

```diff
@@ -159,3 +197,152 @@ func (receiver *DeploymentPlugin) AggregateStatues(
+		}
+	}
+
+	return templateDispatched, nil
+}
+
+func isResourceTemplateSynced(sourceObject, fedObject *unstructured.Unstructured) (bool, error) {
+	synced, err := isResourceTemplateSyncedToFederatedObject(sourceObject, fedObject)
+	if err != nil || !synced {
+		return synced, err
+	}
+
+	return isResourceTemplateSyncedToMemberCluster(fedObject)
+}
+
+func isResourceTemplateSyncedToMemberCluster(fedObject *unstructured.Unstructured) (bool, error) {
+	if fedObject == nil {
+		return false, nil
+	}
+
+	observedGeneration, found, err := unstructured.NestedInt64(fedObject.Object, "status", "syncedGeneration")
+	if err != nil || !found {
+		return false, err
+	}
+
+	// check if the latest resource template has been synced to member clusters
+	return observedGeneration == fedObject.GetGeneration(), nil
+}
+
+func isResourceTemplateSyncedToFederatedObject(sourceObject, fedObject *unstructured.Unstructured) (bool, error) {
+	templateMap, exists, err := unstructured.NestedMap(fedObject.Object, "spec", "template")
+	if err != nil || !exists {
+		return false, err
+	}
+
+	fedAnnotations := fedObject.GetAnnotations()
+	if len(fedAnnotations) == 0 {
+		return false, nil
+	}
+
+	if !areStringMapsSynced(
+		sourceObject.GetAnnotations(),
+		fedAnnotations,
+		fedAnnotations[common.ObservedAnnotationKeysAnnotation],
+	) {
+		return false, nil
+	}
+
+	if !areStringMapsSynced(
+		sourceObject.GetLabels(),
+		fedObject.GetLabels(),
+		fedAnnotations[common.ObservedLabelKeysAnnotation],
+	) {
+		return false, nil
+	}
+
+	prunedSourceObj, err := pruneSourceObj(
+		sourceObject,
+		[]byte(fedAnnotations[common.TemplateGeneratorMergePatchAnnotation]),
+	)
+	if err != nil {
+		return false, err
+	}
+
+	if reflect.DeepEqual(prunedSourceObj, templateMap) {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func areStringMapsSynced(
+	sourceMap, federatedMap map[string]string,
+	observedKeys string,
+) bool {
+	fedKeys, ignoredKeys := parseObservedKeys(observedKeys)
+	allKeys := fedKeys.Union(ignoredKeys)
+
+	if !allKeys.Equal(sets.KeySet(sourceMap)) {
+		return false
+	}
+
+	for key := range fedKeys {
+		if sourceMap[key] != federatedMap[key] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func parseObservedKeys(keys string) (fed, nonFed sets.Set[string]) {
+	keySegments := strings.Split(keys, "|")
+	if len(keySegments) != 2 {
+		return nil, nil
+	}
+
+	if len(keySegments[0]) != 0 {
+		fed = sets.New(strings.Split(keySegments[0], ",")...)
+	}
+
+	if len(keySegments[1]) != 0 {
+		nonFed = sets.New(strings.Split(keySegments[1], ",")...)
+	}
+	return
+}
+
+func pruneSourceObj(sourceObj *unstructured.Unstructured, mergePatch []byte) (map[string]interface{}, error) {
+	sourceJSON, err := sourceObj.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	prunedSourceJSON, err := jsonpatch.MergePatch(
+		sourceJSON,
+		mergePatch,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var ret map[string]any
+	if err := utiljson.Unmarshal(prunedSourceJSON, &ret); err != nil {
```

Comment:
Let's add a comment here documenting why `utiljson` is required and `json` doesn't work.